### PR TITLE
#167313360 Password reset 

### PR DESCRIPTION
**Pull request by Anguandia**
User should be able to reset their password

#### What does this PR do?

  Add password reset feature

#### Tasks to be accomplished

  - Present user with a form to take in email
  - Perform basic input validation
  - Display response from backend with link to email
  - Open password change page on click of the reset link
  - Display password reset response

#### How to manually test this feature

  - Navigate to the reset page by clicking the forgot password link on the login page
  - Enter your valid and registered email and follow instructions
  - Upon success, try logging in with the old, then new password to confirm the password has been reset

#### Related pivotal tracker stories
  [finishes #167313360]

#### Screenshots (if appropriate)

Email input form
image

mobile view
image

email sent response
image

Mobile view
image

reset password page
image

mobile view
image

successful reset response
image

mobile view
image

### CODE REVIEW

**Review by nshutijonathan — DISMISSED** [review]

**Review by habinezadalvan — CHANGES_REQUESTED** [review]
@Anguandia, could you please fix the Travis issue. Otherwise, you have done great work.

File: settings.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"editor.formatOnSave": true}
```
Comment (minega25) [inline settings.json:1]: @mike This file is used by vscode. I think it should be in gitignore for now

File: src/components/passwordreset/components/input.test.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import Input from './input';
+
+describe('test input element component', () => {
+  it('should test the input component', () => {
+    shallow(<Input name="email" type="email" handleChange={() => ''} />);
```
Comment (minega25) [inline src/components/passwordreset/components/input.test.js]: @Anguandia I suggest we expect something to equal some test matcher.

File: src/components/passwordreset/feedbackContainer/feedback.test.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import Feedback from './feedback';
+
+describe('Test feedback component', () => {
+  it('should test the feedback component', () => {
+    shallow(<Feedback message="reset successful" summary="reset success" />);
```
Comment (minega25) [inline src/components/passwordreset/feedbackContainer/feedback.test.js]: same case here.

**Review by nkalyesubula — CHANGES_REQUESTED** [review]
Good work Mike. Please fix failing tests, and also provide screenshots for responsiveness

File: src/redux/actions/actionTypes.js [hunk 4]
```diff
@@ -27,3 +27,8 @@ export const TWITTER_AUTH_FAIL = 'TWITTER_AUTH_FAIL';
 // this will contain actions
 export const CREATE_ARTICLE_SUCCESS = 'CREATE_ARTICLE_SUCCESS';
 export const CREATE_ARTICLE_FAIL = 'CREATE_ARTICLE_FAIL';
+//Reset password actions
+export const ResponseMessage = 'Message';
+export const FormData = 'FormData';
+export const Status = 'Status';
+export const Summary = 'Summary';
```
Comment (minega25) [inline src/redux/actions/actionTypes.js]: @Anguandia As you might have noticed the team is using a standard naming for action types. Kindly look into these.

File: src/redux/reducers/passwordreset/passwordreset.test.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+// import React from 'react';
+// import renderer from 'react-test-renderer';
```
Comment (minega25) [inline src/redux/reducers/passwordreset/passwordreset.test.js]: Kindly delete these lines.

**Review by minega25 — CHANGES_REQUESTED** [review]
Great Job @Anguandia. Please kindly take a look at the below comments and let me know what you think.

File: src/components/passwordreset/feedbackContainer/index.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { connect } from 'react-redux';
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import Feed from './feedback';
+// import Footer from '../footer';
```
Comment (habinezadalvan) [inline src/components/passwordreset/feedbackContainer/index.js]: I think it would be better if you remove these commented lines if you are using them.

**Review by habinezadalvan — CHANGES_REQUESTED** [review]
Codes look good to me. But you have to fix the Travis issue.

File: src/redux/actions/actionTypes.js (hunk 4 above)
Comment (habinezadalvan) [reply src/redux/actions/actionTypes.js]: +1 @minega25

**Comment by Anguandia** [comment]
Hi @all, fixed the travis issue, working on the rest

File: src/components/Routes.js [hunk 7]
```diff
@@ -31,6 +33,8 @@ const Routes = () => (
     <Route exact path="/google/social-login" component={verifyAuth} />
     <Route exact path="/twitter/social-login" component={verifyAuth} />
     <Route exact path="/facebook/social-login" component={verifyAuth} />
+    <Route exact path="/reset" component={ResetRequest} />
+    <Route exact path="/response" component={Feedback} />
```
Comment (SokoPaulSokool) [inline src/components/Routes.js:44]: It would be good if you use a descriptive path

**Review by SokoPaulSokool — CHANGES_REQUESTED** [review]

File: src/redux/reducers/passwordreset/passwordreset.test.js (hunk 5 above)
Comment (Anguandia) [reply src/redux/reducers/passwordreset/passwordreset.test.js]: Deleted

File: src/components/passwordreset/feedbackContainer/index.js (hunk 6 above)
Comment (Anguandia) [reply src/components/passwordreset/feedbackContainer/index.js]: Removed

**Comment by SokoPaulSokool** [comment]
You may also think about adding a progress bar when someone clicks send

File: src/components/Routes.js (hunk 7 above)
Comment (Anguandia) [reply src/components/Routes.js:44]: @SokoPaulSokool ,
Paths renamed to resetpassword and resetpasswordresponse respectively.
A small modification though will be needed in the backend link sent in email.
Will raise it with @salviosage

File: settings.json (hunk 1 above)
Comment (Anguandia) [reply settings.json:1]: @minega25, settings.json added to git ignore

File: src/components/passwordreset/components/input.test.js (hunk 2 above)
Comment (Anguandia) [reply src/components/passwordreset/components/input.test.js]: @minega25 , changed the individual component tests to check specific component element properties, thanks for the observation

File: src/components/passwordreset/feedbackContainer/feedback.test.js (hunk 3 above)
Comment (Anguandia) [reply src/components/passwordreset/feedbackContainer/feedback.test.js]: --same as above--

File: src/redux/actions/actionTypes.js (hunk 4 above)
Comment (Anguandia) [reply src/redux/actions/actionTypes.js]: @minega25 , @habinezadalvan , actiontype naming streamlined, thanks

File: src/redux/reducers/passwordreset/passwordreset.test.js (hunk 5 above)
Comment (Anguandia) [reply src/redux/reducers/passwordreset/passwordreset.test.js]: All commented-out code deleted

File: src/components/Routes.js (hunk 7 above)
Comment (Anguandia) [reply src/components/Routes.js:44]: @SokoPaulSokool, a simple growing-ribbon progress bar added to the top of the page, growing from    0-length, red color to full-length green every 2 seconds while still processing data

Comment (Anguandia) [reply src/components/Routes.js:44]: Also added mobile view pictures

File: src/components/passwordreset/components/navbar.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  <Navbar collapseOnSelect expand="lg" bg="transparent" variant="light" sticky="top" className="nav">
+    <Navbar.Brand>
+      {' '}
+      <Logo />
+      {' '}
+    </Navbar.Brand>
+    <Navbar.Toggle aria-controls="responsive-navbar-nav" className="nav" />
+    <Navbar.Collapse className="justify-content-end">
+      <Nav.Link href="/articles">
+Articles
```
Comment (SokoPaulSokool) [inline src/components/passwordreset/components/navbar.js]: We already have a navbar component in the shared components so there is no need for a new one

File: src/redux/actions/passwordreset/index.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+      method,
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(userData),
+      mode: 'cors',
+    },
+  );
+};
```
Comment (SokoPaulSokool) [inline src/redux/actions/passwordreset/index.js]: We are not following this pattern of writing actions. Look at the way your teammates have been implementing and adjust accordingly.

File: src/components/passwordreset/index.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+const mapState = (state) => ({
+  message: state.passwordReset.message,
+  email: state.passwordReset.email,
+});
+
+const dispatchProps = (dispatch) => ({
+  responseMessage: (m, n) => dispatch(Message(m, n)),
+});
```
Comment (SokoPaulSokool) [inline src/components/passwordreset/index.js]: This is not the way we have been implementing dispatch. Adjust and follow the pattern we have been using.

File: src/components/passwordreset/components/button.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+/**
+ * @description - Button Component
+ * @param {object} props
+ * @returns {JSX} - Button JSX template
+ */
+const Button = ({ value }) => (
```
Comment (SokoPaulSokool) [inline src/components/passwordreset/components/button.js]: Your button does not have effects like other buttons. I think you don't need to create a new button, there is a submitButton component

File: src/components/passwordreset/components/input.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import propTypes from 'prop-types';
+
+/**
+ * @description - Form input element Component
+ * @param {object} props
+ * @returns {JSX} - Form minput JSX template
+ */
+const Input = ({
```
Comment (SokoPaulSokool) [inline src/components/passwordreset/components/input.js]: We already have a textInput component there is no need for creating a new one

**Review by SokoPaulSokool — CHANGES_REQUESTED** [review]
Good job. In the future designing a component for something you already saw, first find out how it was implemented. You will in most times realize that it was already implemented and it will save you time.

And when working on a team project, if there is a workflow like redux, try to look at other teammates' work and see if there is a pattern in the implementation and follow it. If you do things on your own, they may be right but may disrupt the organization of the project code flow.

File: src/components/passwordreset/components/navbar.js (hunk 8 above)
Comment (Anguandia) [reply src/components/passwordreset/components/navbar.js]: Adopted existing common navbar.

**Review by minega25 — APPROVED** [review]
LGTM.

**Review by habinezadalvan — CHANGES_REQUESTED** [review]
I'm impressed by the work you've done so far. However, I would like you to link up the forgot password link on landing page with your work. Again, I would like to know why the reset password link I receive from the email doesn't work. Thank. Otherwise, you did a great job.

File: src/components/login/index.js [hunk 13]
```diff
@@ -100,7 +100,7 @@ export class Login extends Component {
                 <TextInput type="password" label="Password" name="password" value={password} onChange={this.onChange} error={error} />
                 <SubmitButton value="Login" />
               </form>
-              <a href="/resetrequest">Forgot password ?</a>
+              <a href="/reset">Forgot password ?</a>
```
Comment (salviosage) [inline src/components/login/index.js:103]: please do change this to `/resetpassword` as it is in the routes file

File: src/components/passwordreset/components/input.js (hunk 12 above)
Comment (Anguandia) [reply src/components/passwordreset/components/input.js]: Now using the existing text input

File: src/components/passwordreset/components/button.js (hunk 11 above)
Comment (Anguandia) [reply src/components/passwordreset/components/button.js]: Switched to using the existing button

File: src/components/passwordreset/index.js (hunk 10 above)
Comment (Anguandia) [reply src/components/passwordreset/index.js]: Adopted the way used by the team

File: src/redux/actions/passwordreset/index.js (hunk 9 above)
Comment (Anguandia) [reply src/redux/actions/passwordreset/index.js]: Modified actions file

File: src/components/login/index.js (hunk 13 above)
Comment (Anguandia) [reply src/components/login/index.js:103]: Thanks, changed

**Comment by Anguandia** [comment]
Thank you for pointing that out @habinezadalvan, the TTL requested me to change the pathname to a more descriptive name, so I changed it to resetpassword. I have however addressed that issue, the router looks for a match of reset in the path. thanks again for pointing out, now you can retry

**Review by habinezadalvan — APPROVED** [review]
LGTM

**Pull request merged**